### PR TITLE
Add ALPN extension to the server side

### PR DESCRIPTION
**Pull request by xkqian**
This will add alpn extension support the server side.
Also make the exist alpn cases available for both tls1.2 and tls1.3

### CODE REVIEW

File: library/ssl_misc.h [hunk 1]
```diff
@@ -2280,4 +2280,16 @@ int mbedtls_ssl_tls13_get_sig_alg_from_pk( mbedtls_ssl_context *ssl,
...
+#if defined(MBEDTLS_SSL_ALPN)
+int mbedtls_ssl_parse_alpn_ext( mbedtls_ssl_context *ssl,
+                                const unsigned char *buf,
+                                const unsigned char *end );
+
+
+int mbedtls_ssl_write_alpn_ext( mbedtls_ssl_context *ssl,
+                                unsigned char *buf,
+                                unsigned char *end,
+                                size_t *olen );
```
Comment (yuhaoth) [inline library/ssl_misc.h:2292]:
```suggestion
                                size_t *out_len );
```

File: library/ssl_tls.c [hunk 2]
```diff
@@ -8285,4 +8285,104 @@ int mbedtls_ssl_parse_server_name_ext( mbedtls_ssl_context *ssl,
...
+    if( ssl->conf->alpn_list == NULL )
+        return( 0 );
+
+    /*
+     * opaque ProtocolName<1..2^8-1>;
+     *
+     * struct {
+     *     ProtocolName protocol_name_list<2..2^16-1>
+     * } ProtocolNameList;
+     */
```
Comment (yuhaoth) [inline library/ssl_tls.c:8309]:
```suggestion
    /*
     * RFC7301, section 3.1
     *   opaque ProtocolName<1..2^8-1>;
     *
     *   struct {
     *       ProtocolName protocol_name_list<2..2^16-1>
     *   } ProtocolNameList;
     */
```

File: library/ssl_tls.c [hunk 3]
```diff
@@ -8285,4 +8285,104 @@ int mbedtls_ssl_parse_server_name_ext( mbedtls_ssl_context *ssl,
...
+
+    /* Min length is 2 ( list_len ) + 1 ( name_len ) + 1 ( name ) */
```
Comment (yuhaoth) [inline library/ssl_tls.c:8311]:
```suggestion
    /* 
     * protocal_name_list_len    2 bytes
     * protocal_name_len          1 bytes
     * protocal_name                 >=1 byte
     */
```

File: library/ssl_tls.c [hunk 4]
```diff
@@ -8285,4 +8285,104 @@ int mbedtls_ssl_parse_server_name_ext( mbedtls_ssl_context *ssl,
 }
 #endif /* MBEDTLS_SSL_SERVER_NAME_INDICATION */
 
+#if defined(MBEDTLS_SSL_ALPN)
+int mbedtls_ssl_parse_alpn_ext( mbedtls_ssl_context *ssl,
+                                const unsigned char *buf,
+                                const unsigned char *end )
+{
+    const unsigned char *p = buf;
+    size_t list_len;
```
Comment (yuhaoth) [inline library/ssl_tls.c:8294]:
```suggestion
    size_t protocol_name_list_len;
```

File: library/ssl_tls.c [hunk 5]
```diff
@@ -8285,4 +8285,104 @@ int mbedtls_ssl_parse_server_name_ext( mbedtls_ssl_context *ssl,
...
+
+    const unsigned char *cur_alpn;
+    size_t cur_alpn_len;
```
Comment (yuhaoth) [inline library/ssl_tls.c:8297]:
```suggestion
    const unsigned char *protocol_name;
    size_t protocol_name_len;
```

File: library/ssl_tls.c [hunk 6]
```diff
@@ -8285,4 +8285,104 @@ int mbedtls_ssl_parse_server_name_ext( mbedtls_ssl_context *ssl,
...
+    MBEDTLS_SSL_PEND_FATAL_ALERT(
+            MBEDTLS_SSL_ALERT_MSG_NO_APPLICATION_PROTOCOL,
+            MBEDTLS_ERR_SSL_NO_APPLICATION_PROTOCOL );
+    return( MBEDTLS_ERR_SSL_NO_APPLICATION_PROTOCOL );
+}
+
+int mbedtls_ssl_write_alpn_ext( mbedtls_ssl_context *ssl,
+                                unsigned char *buf,
+                                unsigned char *end,
+                                size_t *olen )
```
Comment (yuhaoth) [inline library/ssl_tls.c:8354]:
```suggestion
                                size_t *out_len )
```

File: tests/ssl-opt.sh [hunk 7]
```diff
@@ -10517,6 +10510,32 @@ run_test    "TLS 1.3: alpn - gnutls" \
...
+            0 \
+            -s "server state: MBEDTLS_SSL_HELLO_REQUEST" \
+            -s "server state: MBEDTLS_SSL_SERVER_HELLO" \
+            -s "server state: MBEDTLS_SSL_ENCRYPTED_EXTENSIONS" \
+            -s "server state: MBEDTLS_SSL_CERTIFICATE_REQUEST" \
+            -s "server state: MBEDTLS_SSL_SERVER_CERTIFICATE" \
+            -s "server state: MBEDTLS_SSL_CERTIFICATE_VERIFY" \
+            -s "server state: MBEDTLS_SSL_SERVER_FINISHED" \
+            -s "server state: MBEDTLS_SSL_CLIENT_FINISHED" \
+            -s "server state: MBEDTLS_SSL_HANDSHAKE_WRAPUP" \
```
Comment (yuhaoth) [inline tests/ssl-opt.sh:10531]:
```suggestion
```

**Review by yuhaoth — CHANGES_REQUESTED** [review]
Some improvement and GnuTLS test is missing

**Review by yuhaoth — DISMISSED** [review]
LGTM.

File: library/ssl_tls13_server.c [hunk 8]
```diff
@@ -1372,6 +1396,15 @@ static int ssl_tls13_write_encrypted_extensions_body( mbedtls_ssl_context *ssl,
...
 
     ((void) ssl);
+    ((void) ret);
+    ((void) output_len);
+
+#if defined(MBEDTLS_SSL_ALPN)
+    ret = mbedtls_ssl_write_alpn_ext( ssl, p, end, &output_len );
+    if( ret != 0 )
+        return( ret );
+    p  += output_len;
```
Comment (ronald-cron-arm) [inline library/ssl_tls13_server.c:1406]:
```suggestion
    p += output_len;
```

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
This looks quite good to me, just some minor suggestion for improvement.

File: library/ssl_tls.c [hunk 9]
```diff
@@ -8285,4 +8285,111 @@ int mbedtls_ssl_parse_server_name_ext( mbedtls_ssl_context *ssl,
...
+     * 4 . 5    protocol list length
+     * 6 . 6    protocol name length
+     * 7 . 7+n  protocol name
+     */
+    MBEDTLS_PUT_UINT16_BE( MBEDTLS_TLS_EXT_ALPN, p, 0 );
+
+    *out_len = 7 + strlen( ssl->alpn_chosen );
+
+    MBEDTLS_PUT_UINT16_BE( *out_len - 4, p, 2 );
+    MBEDTLS_PUT_UINT16_BE( *out_len - 6, p, 4 );
```
[the remark this answers is not in the document]
Comment (xkqian) [reply library/ssl_tls.c:8386]: MBEDTLS_PUT_UINT16_BE( protocol_name_len + 1, p, 4 );

File: library/ssl_tls.c [hunk 10]
```diff
@@ -8285,4 +8285,122 @@ int mbedtls_ssl_parse_server_name_ext( mbedtls_ssl_context *ssl,
...
+     * RFC7301, section 3.1
+     *      opaque ProtocolName<1..2^8-1>;
+     *
+     *      struct {
+     *          ProtocolName protocol_name_list<2..2^16-1>
+     *      } ProtocolNameList;
+     */
+
+    /*
+     * protocal_name_list_len    2 bytes
```
Comment (ronald-cron-arm) [inline library/ssl_tls.c:8313]:
```suggestion
     * protocol_name_list_len    2 bytes
```

File: library/ssl_tls.c [hunk 11]
```diff
@@ -8285,4 +8285,122 @@ int mbedtls_ssl_parse_server_name_ext( mbedtls_ssl_context *ssl,
...
+     * protocal_name_len         1 bytes
```
Comment (ronald-cron-arm) [inline library/ssl_tls.c:8314]:
```suggestion
     * protocol_name_len         1 bytes
```

File: library/ssl_tls.c [hunk 12]
```diff
@@ -8285,4 +8285,122 @@ int mbedtls_ssl_parse_server_name_ext( mbedtls_ssl_context *ssl,
...
+     * protocal_name             >=1 byte
```
Comment (ronald-cron-arm) [inline library/ssl_tls.c:8315]:
```suggestion
     * protocol_name             >=1 byte
```

File: library/ssl_tls.c [hunk 13]
```diff
@@ -8285,4 +8285,122 @@ int mbedtls_ssl_parse_server_name_ext( mbedtls_ssl_context *ssl,
...
+     * 6 . 6    protocol name length
+     * 7 . 7+n  protocol name
+     */
+    MBEDTLS_PUT_UINT16_BE( MBEDTLS_TLS_EXT_ALPN, p, 0 );
+
+    *out_len = 7 + protocol_name_len;
+
+    MBEDTLS_PUT_UINT16_BE( protocol_name_len + 3, p, 2 );
+    MBEDTLS_PUT_UINT16_BE( protocol_name_len + 1, p, 4 );
+    p[6] = MBEDTLS_BYTE_0( protocol_name_len );
```
Comment (ronald-cron-arm) [inline library/ssl_tls.c:8399]:
```suggestion
    /* Note: the length of the chosen protocol has been checked to be less
     * than 255 bytes in `mbedtls_ssl_conf_alpn_protocols`.
     */
    p[6] = MBEDTLS_BYTE_0( protocol_name_len );
```

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
Final very minor things. Otherwise, this looks good to me.

**Review by ronald-cron-arm — APPROVED** [review]
LGTM, @yuhaoth please have a look to the last version.

**Review by yuhaoth — APPROVED** [review]
LGTM

**Pull request merged**